### PR TITLE
[oceanbase] Add new deserialization schema with runtime converter

### LINKED ISSUE #980: [oceanbase] Use customized runtime converter to replace JdbcValueConverters
Implement a runtime converter to make the logic of type conversion clearer, which may improve accuracy and performance.

### LINKED ISSUE #1289: [oceanbase] add support for `JSON` type
**Is your feature request related to a problem? Please describe.**
No.

**Describe the solution you'd like**
OceanBase CE supports JSON type from 3.1.3, we should update the connector for it.

**Describe alternatives you've considered**
Add JSON type in type converter.

### PR DESCRIPTION
**Pull request by whhe**
The previous logic is to do the first conversion with OceanBaseJdbcConverter helper methods, and then use runtime converters in JdbcValueConverters to do the second conversion. Now we need only do conversion one time with the new deserialization schema.

close #980 close #1289

### CODE REVIEW

**Comment by whhe** [comment]
Any maintainer can help review this? @GOODBOY008 @Jiabao-Sun

**Comment by leonardBang** [comment]
@ruanhang1993 Would you like to help review this PR ?

**Comment by whhe** [comment]
There might be something wrong on the self-hosted agent machine, the oceanbase container can't be started now. @leonardBang

**Comment by leonardBang** [comment]
> There might be something wrong on the self-hosted agent machine, the oceanbase container can't be started now. @leonardBang

let me see see

File: flink-connector-oceanbase-cdc/src/main/java/com/ververica/cdc/connectors/oceanbase/source/RowDataOceanBaseDeserializationSchema.java [hunk 1]
```diff
@@ -0,0 +1,679 @@
...
+            public Object convertSnapshotEvent(Object object) {
+                if (object instanceof byte[]) {
+                    return "1".equals(new String((byte[]) object, StandardCharsets.UTF_8));
+                }
+                return Boolean.parseBoolean(object.toString()) || "1".equals(object.toString());
+            }
+
+            @Override
+            public Object convertChangeEvent(String string) {
+                return "1".equals(string);
```
Comment (ruanhang1993) [inline flink-connector-oceanbase-cdc/src/main/java/com/ververica/cdc/connectors/oceanbase/source/RowDataOceanBaseDeserializationSchema.java:309]: Why not contain `"true".equals(string)` here?

File: flink-connector-oceanbase-cdc/src/main/java/com/ververica/cdc/connectors/oceanbase/source/RowDataOceanBaseDeserializationSchema.java [hunk 2]
```diff
@@ -0,0 +1,679 @@
...
+            case BINARY:
+                return convertToBinary();
+            case VARBINARY:
+                return convertToBytes();
+            case DECIMAL:
+                return createDecimalConverter((DecimalType) type);
+            case ARRAY:
+                return createArrayConverter();
+            default:
+                return null;
```
Comment (ruanhang1993) [inline flink-connector-oceanbase-cdc/src/main/java/com/ververica/cdc/connectors/oceanbase/source/RowDataOceanBaseDeserializationSchema.java:252]: There may need to throw an `UnsupportedOperationException`.

File: flink-connector-oceanbase-cdc/src/test/resources/ddl/column_type_test.sql [hunk 3]
```diff
@@ -40,11 +43,12 @@ CREATE TABLE full_types
     big_decimal_c DECIMAL(65, 1),
     date_c        DATE,
     time_c        TIME(0),
-    datetime3_c   TIMESTAMP(3),
-    datetime6_c   TIMESTAMP(6),
```
Comment (ruanhang1993) [inline flink-connector-oceanbase-cdc/src/test/resources/ddl/column_type_test.sql:44]: We also need to add a column whose type is `TIMESTAMP(p)`.

File: flink-connector-oceanbase-cdc/src/test/java/com/ververica/cdc/connectors/oceanbase/table/OceanBaseConnectorITCase.java [hunk 4]
```diff
@@ -382,15 +393,17 @@ public void testAllDataTypes() throws Exception {
                         + "    timestamp_c TIMESTAMP,\n"
```
Comment (ruanhang1993) [inline flink-connector-oceanbase-cdc/src/test/java/com/ververica/cdc/connectors/oceanbase/table/OceanBaseConnectorITCase.java:393]: The type of `timestamp_c` should be `TIMESTAMP_LTZ`, if I don't misunderstand it.

**Review by ruanhang1993 — COMMENTED** [review]
Thanks for your work. I have left some comments.
Besides these, I find that there are many duplicated code between the new code and the code in module `flink-connector-debezium`, e.g. `AppendMetadataCollector`. Is there a way to reuse them?

File: flink-connector-oceanbase-cdc/src/main/java/com/ververica/cdc/connectors/oceanbase/source/RowDataOceanBaseDeserializationSchema.java [hunk 5]
```diff
@@ -0,0 +1,679 @@
...
+        return new RowDataOceanBaseDeserializationSchema.Builder();
+    }
+
+    RowDataOceanBaseDeserializationSchema(
+            RowType physicalDataType,
+            OceanBaseMetadataConverter[] metadataConverters,
+            TypeInformation<RowData> resultTypeInfo,
+            ZoneId serverTimeZone) {
+        this.hasMetadata = checkNotNull(metadataConverters).length > 0;
+        this.appendMetadataCollector = new OceanBaseAppendMetadataCollector(metadataConverters);
```
[the remark this answers is not in the document]
Comment (whhe) [reply flink-connector-oceanbase-cdc/src/main/java/com/ververica/cdc/connectors/oceanbase/source/RowDataOceanBaseDeserializationSchema.java:89]: Can you tell me the reason? Here it's already passed as `RowType` in the constructor parameters.

File: flink-connector-oceanbase-cdc/src/main/java/com/ververica/cdc/connectors/oceanbase/source/RowDataOceanBaseDeserializationSchema.java [hunk 6]
```diff
@@ -0,0 +1,679 @@
...
+            private static final long serialVersionUID = 1L;
+
+            @Override
+            public Object convert(Object object) throws Exception {
+                int arity = fieldNames.length;
+                GenericRowData row = new GenericRowData(arity);
+                Map<String, Object> fieldMap = (Map<String, Object>) object;
+                for (int i = 0; i < arity; i++) {
+                    String fieldName = fieldNames[i];
+                    row.setField(i, fieldConverters[i].convert(fieldMap.get(fieldName)));
```
[the remark this answers is not in the document]
Comment (whhe) [reply flink-connector-oceanbase-cdc/src/main/java/com/ververica/cdc/connectors/oceanbase/source/RowDataOceanBaseDeserializationSchema.java:275]: It's possible and acceptable to ruturn null here.

File: flink-connector-oceanbase-cdc/src/test/resources/ddl/column_type_test.sql (hunk 3 above)
Comment (whhe) [reply flink-connector-oceanbase-cdc/src/test/resources/ddl/column_type_test.sql:44]: OK.

File: flink-connector-oceanbase-cdc/src/test/java/com/ververica/cdc/connectors/oceanbase/table/OceanBaseConnectorITCase.java (hunk 4 above)
Comment (whhe) [reply flink-connector-oceanbase-cdc/src/test/java/com/ververica/cdc/connectors/oceanbase/table/OceanBaseConnectorITCase.java:393]: You're right. Fixed.

File: flink-connector-oceanbase-cdc/src/main/java/com/ververica/cdc/connectors/oceanbase/source/RowDataOceanBaseDeserializationSchema.java (hunk 2 above)
Comment (whhe) [reply flink-connector-oceanbase-cdc/src/main/java/com/ververica/cdc/connectors/oceanbase/source/RowDataOceanBaseDeserializationSchema.java:252]: OK.

File: flink-connector-oceanbase-cdc/src/test/java/com/ververica/cdc/connectors/oceanbase/table/OceanBaseConnectorITCase.java (hunk 4 above)
Comment (whhe) [reply flink-connector-oceanbase-cdc/src/test/java/com/ververica/cdc/connectors/oceanbase/table/OceanBaseConnectorITCase.java:393]: I tried `TIMESTAMP_LTZ` here, and the timezone config for `tEnv` seems not work. Maybe it's not supported in `values` connector? Keep it for now.

**Comment by whhe** [comment]
> Thanks for your work. I have left some comments. Besides these, I find that there are many duplicated code between the new code and the code in module `flink-connector-debezium`, e.g. `AppendMetadataCollector`. Is there a way to reuse them?

It's hard to reuse codes in `RowDataDebeziumDeserializeSchema` and `AppendMetadataCollector` in debezium module, because they both are final classes.

File: flink-connector-oceanbase-cdc/src/main/java/com/ververica/cdc/connectors/oceanbase/source/RowDataOceanBaseDeserializationSchema.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (ruanhang1993) [reply flink-connector-oceanbase-cdc/src/main/java/com/ververica/cdc/connectors/oceanbase/source/RowDataOceanBaseDeserializationSchema.java:89]: My bad. It is fine here.

**Comment by whhe** [comment]
@ruanhang1993 Any question for now?

**Comment by ruanhang1993** [comment]
> @ruanhang1993 Any question for now?

No more questions from my side. : )

**Comment by whhe** [comment]
Any progress for this PR? @leonardBang

**Review by leonardBang — APPROVED** [review]
Thanks @whhe and @ruanhang1993 for the great work, LGTM

**Pull request merged**